lagrangebench/evaluate/rollout.py

```diff
@@ -54,64 +102,80 @@ def eval_single_rollout(
+    batch_size, n_nodes_max, _, dim = pos_input_batch.shape
+
     # if n_rollout_steps set to -1, use the whole trajectory
-    if n_rollout_steps < 0:
-        n_rollout_steps = pos_input.shape[1] - t_window
+    if n_rollout_steps == -1:
+        n_rollout_steps = pos_input_batch.shape[2] - t_window
 
-    initial_positions = pos_input[:, 0:t_window]  # (n_nodes, t_window, dim)
-    traj_len = n_rollout_steps + n_extrap_steps  # (n_nodes, traj_len - t_window, dim)
-    ground_truth_positions = pos_input[:, t_window : t_window + traj_len]
-    current_positions = initial_positions  # (n_nodes, t_window, dim)
-    n_nodes, _, dim = ground_truth_positions.shape
+    current_positions_batch = pos_input_batch[:, :, 0:t_window]
+    # (batch, n_nodes, t_window, dim)
+    traj_len = n_rollout_steps + n_extrap_steps
+    target_positions_batch = pos_input_batch[:, :, t_window : t_window + traj_len]
 
-    predictions = jnp.zeros((traj_len, n_nodes, dim))
+    predictions_batch = jnp.zeros((batch_size, traj_len, n_nodes_max, dim))
+    neighbors_batch = broadcast_to_batch(neighbors, batch_size)
+    preprocess_eval_vmap = vmap(case.preprocess_eval, in_axes=(0, 0))
+
+    forward_eval = partial(
+        _forward_eval,
+        model_apply=model_apply,
+        case_integrate=case.integrate,
+    )
+    forward_eval_vmap = vmap(forward_eval, in_axes=(None, None, 0, 0, 0))
 
     step = 0
     while step < n_rollout_steps + n_extrap_steps:
-        sample = (current_positions, particle_type)
-        features, neighbors = case.preprocess_eval(sample, neighbors)
+        sample_batch = (current_positions_batch, particle_type_batch)
 
-        if neighbors.did_buffer_overflow is True:
-            edges_ = neighbors.idx.shape
-            print(f"(eval) Reallocate neighbors list {edges_} at step {step}")
-            _, neighbors = case.allocate_eval(sample)
-            print(f"(eval) To list {neighbors.idx.shape}")
+        # 1. preprocess features
+        features_batch, neighbors_batch = preprocess_eval_vmap(
+            sample_batch, neighbors_batch
+        )
 
-            continue
+        # 2. check whether list overflowed and fix it if so
+        if neighbors_batch.did_buffer_overflow.sum() > 0:
+            # check if the neighbor list is too small for any of the samples
+            # if so, reallocate the neighbor list
 
-        # predict
-        pred, _ = model_apply(params, state, (features, particle_type))
+            print(f"(eval) Reallocate neighbors list at step {step}")
+            ind = jnp.argmax(neighbors_batch.did_buffer_overflow)
+            sample = broadcast_from_batch(sample_batch, index=ind)
 
-        next_position = case.integrate(pred, current_positions)
+            _, nbrs_temp = case.allocate_eval(sample)
+            print(
+                f"(eval) From {neighbors_batch.idx[ind].shape} to {nbrs_temp.idx.shape}"
+            )
+            neighbors_batch = broadcast_to_batch(nbrs_temp, batch_size)
```

Comment:
what happens here if two batches have a neighbor overflow at the same step?

Reply:
`jnp.argmax` returns the index of the first max value, if there are multiple equal max values. Here, this means the first overflown neighbor list.

Reply:
But by doing this don't we only reallocate the first overflown list?

Reply:
Yes, we reallocate based on the first overflown neighbor list. But given that reallocation happens rarely, this should be fine. The normal case would be: we reallocate based on the first overflown list, then another overflown list still needs reallocation because it needs even more space, but then we also fix that second one. 

I guess you mean if the overflow is based on different criteria, e.g. individual cell capacity vs edge list capacity. I guess the only solution to that is to have a reasonable `neighbor_list_multiplier`, say >=1.25. In general, larger `cell_capacity` means clustering, and clustering means more edges, thus a larger required edge list. If we get this property right, the rest of the batch should fall within a 25% larger edge list.